File: 2019/day7/day7.py
f = open('input', 'r')

inst = f.read().split(',')
out = 0

def proc(tape, ph, inp, index):
    #global out
    global ready
    global reg
    pos = 0
    phase = True
    while not tape[pos] == '99':
        if tape[pos][-1] == '1': #add
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            add = int(x) + int(y)
            tape[int(tape[pos+3])] = str(add)
            pos += 4
        elif tape[pos][-1] == '2': #mul
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            prod = int(x) * int(y)
            tape[int(tape[pos+3])] = str(prod)
            pos += 4
        elif tape[pos] == '3': #read
            i = int(tape[pos+1])
            if phase:
                tape[i] = str(ph)
                phase = False
                pos += 2
            elif ready[index]:
                tape[i] = str(reg[index])
                ready[index] = False
                pos += 2
            #else:
                #tape[i] = str(inp)
            #pos += 2
        elif tape[pos][-1] == '4': #output
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                o = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                o = tape[pos+1]
            print(o)
            #out = int(o)
            reg[(index+1)%5] = int(o)
            ready[(index+1)%5] = True
            pos += 2
        elif tape[pos][-1] == '5': #jtrue
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            if not int(x) == 0:
                pos = int(y)
            else:
                pos += 3
        elif tape[pos][-1] == '6': #jfalse
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            if int(x) == 0:
                pos = int(y)
            else:
                pos += 3
        elif tape[pos][-1] == '7': #lessthan
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            if int(x) < int(y):
                tape[int(tape[pos+3])] = '1'
            else:
                tape[int(tape[pos+3])] = '0'
            pos += 4
        elif tape[pos][-1] == '8': #equals
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            if int(x) == int(y):
                tape[int(tape[pos+3])] = '1'
            else:
                tape[int(tape[pos+3])] = '0'
            pos += 4
        else:
            print('error')
            return -1
    return tape[0]

def permutation(lst):
    if len(lst) == 0:
        return []
    if len(lst) == 1:
        return [lst]

    res = []
    for i in range(len(lst)):
        cur = lst[i]
        remlst = lst[:i] + lst[i+1:]
        for p in permutation(remlst):
            res.append([cur] + p)
    return res

mx = 0
result = []
# shift = [1,0,4,3,2]

# for p in permutation(shift):
#     for i in p:
#         n = inst[:]
#         proc(n, i, out)
#     if out > mx:
#         mx = out
#         result = p
#     out = 0

# print(mx)
# print(result)

mem = []
reg = [0] * 5
ready = [False] * 5
for i in range(5):
    mem += [inst[:]]

File: 2019/day7/day7.cpp
#include <stdio.h>
#include <algorithm>
#include <tuple>
#include <utility>
#include <vector>

std::pair<int, int> getparams(std::vector<int> mem, int pos)
{
    int x, y, inst = mem[pos];

    // get x
    int xmode = (inst / 100) % 10;
    if (xmode == 0) // position
    {
        x = mem[mem[pos + 1]];
    }
    else if (xmode == 1) // immediate
    {
        x = mem[pos + 1];
    }
    else // error
    {
        printf("getparam: x mode error: %d\n", xmode);
        return std::make_pair(-1, -1);
    }

    // only 1 param
    if (inst % 10 == 3 || inst % 10 == 4)
        return std::make_pair(x, 0);

    // get y
    int ymode = (inst / 1000) % 10;
    if (ymode == 0) // position
    {
        y = mem[mem[pos + 2]];
    }
    else if (ymode == 1) // immediate
    {
        y = mem[pos + 2];
    }
    else // error
    {
        printf("getparam: y mode error: %d\n", ymode);
        return std::make_pair(-1, -1);
    }

    return std::make_pair(x, y);
}

int proc(std::vector<std::vector<int>> mem, std::vector<int> shift)
{
    std::vector<bool> done = { false, false, false, false, false };
    std::vector<bool> ready = { true, false, false, false, false };
    int index = 0;
    std::vector<int> reg = { 0, 0, 0, 0, 0 };
    std::vector<int> pos = { 0, 0, 0, 0, 0 };
    std::vector<bool> phase = { true, true, true, true, true };
    while (std::find(done.begin(), done.end(), false) != done.end())
    {
        int inst = mem[index][pos[index]];
        switch (inst % 10)
        {
            case 1: // add
            {
                int x, y;
                std::tie(x, y) = getparams(mem[index], pos[index]);
                mem[index][mem[index][pos[index] + 3]] = x + y;
                pos[index] += 4;
                break;
            }
            case 2: // mul
            {
                int x, y;
                std::tie(x, y) = getparams(mem[index], pos[index]);
                mem[index][mem[index][pos[index] + 3]] = x * y;
                pos[index] += 4;
                break;
            }
            case 3: // input
            {
                if (phase[index])
                {
                    mem[index][mem[index][pos[index] + 1]] = shift[index];
                    phase[index] = false;
                    pos[index] += 2;
                }
                else if (ready[index])
                {
                    mem[index][mem[index][pos[index] + 1]] = reg[index];
                    ready[index] = false;
                    pos[index] += 2;
                }
                else
                {
                    index = (index + 1) % 5;
                }
                break;
            }
            case 4: // output
            {
                int o;
                std::tie(o, std::ignore) = getparams(mem[index], pos[index]);
                reg[(index + 1) % 5] = o;
                ready[(index + 1) % 5] = true;
                pos[index] += 2;
                break;
            }
            case 5: // jtrue
            {
                int x, y;
                std::tie(x, y) = getparams(mem[index], pos[index]);
                if (x != 0)
                    pos[index] = y;
                else
                    pos[index] += 3;
                break;
            }
            case 6: // jfalse
            {
                int x, y;
                std::tie(x, y) = getparams(mem[index], pos[index]);
                if (x == 0)
                    pos[index] = y;
                else
                    pos[index] += 3;
                break;
            }
            case 7: // lessthan
            {
                int x, y;
                std::tie(x, y) = getparams(mem[index], pos[index]);
                if (x < y)
                    mem[index][mem[index][pos[index] + 3]] = 1;
                else
                    mem[index][mem[index][pos[index] + 3]] = 0;
                pos[index] += 4;
                break;
            }
            case 8: // equals
            {
                int x, y;
                std::tie(x, y) = getparams(mem[index], pos[index]);
                if (x == y)
                    mem[index][mem[index][pos[index] + 3]] = 1;
                else
                    mem[index][mem[index][pos[index] + 3]] = 0;
                pos[index] += 4;
                break;
            }
            case 9:
            {
                if (inst == 99)
                {
                    done[index] = true;
                    index = (index + 1) % 5;
                }
                else
                {
                    printf("instruction error: %d\n", inst);
                    return -1;
                }
                break;
            }
            default:
            {
                printf("instruction error: %d\n", inst);
                return -1;
            }
        }
    }

    return reg[0];
}

int main()
{
    std::vector<std::vector<int>> mem;
    for (int i = 0; i < 5; i++)
    {
        mem.push_back(std::vector<int>());
    }

    for (int i = 0; i < 511; i++)
    {
        int p;
        scanf("%d,", &p);
        for (int i = 0; i < 5; i++)
        {
            mem[i].push_back(p);
        }
    }

    // std::vector<int> shift = { 0, 1, 2, 3, 4 }; // part 1
    std::vector<int> shift = { 5, 6, 7, 8, 9 }; // part 2

    int max = 0;
    do
    {
        int out = proc(mem, shift);
        if (out > max)
        {
            max = out;
        }
    } while (std::next_permutation(shift.begin(), shift.end()));

    printf("%d\n", max);
}